### PR TITLE
Sliding Sync: Make `PerConnectionState` immutable

### PR DESCRIPTION
**Pull request by erikjohnston**
This is so that we can cache it.

We also move the sliding sync types to `synapse/types/handlers/sliding_sync.py`. This is mainly in-prep for #17599 to avoid circular imports.

The only change in behaviour is that `RoomSyncConfig.combine_sync_config(..)` now returns a new room sync config rather than mutating in-place.

Reviewable commit-by-commit.

### CODE REVIEW

File: scripts-dev/mypy_synapse_plugin.py [hunk 1]
```diff
@@ -318,6 +320,9 @@ def is_cacheable(
             else:
                 return False, "non-frozen attrs class"
 
+        elif rt.type.is_enum:
```
Comment (MadLittleMods) [inline scripts-dev/mypy_synapse_plugin.py:323]: If we're assuming, could we add `builtins.enum` to `IMMUTABLE_VALUE_TYPES`?

File: scripts-dev/mypy_synapse_plugin.py [hunk 2]
```diff
@@ -318,6 +320,9 @@ def is_cacheable(
...
+            # We assume Enum values are immutable
```
Comment (MadLittleMods) [inline scripts-dev/mypy_synapse_plugin.py:324]: We could check if the enum member values are all immutable to be sure.

File: synapse/types/handlers/sliding_sync.py [hunk 3]
```diff
@@ -499,7 +500,7 @@ def from_room_config(
 
     def deep_copy(self) -> "RoomSyncConfig":
         required_state_map: Dict[str, Set[str]] = {
-            state_type: state_key_set.copy()
+            state_type: set(state_key_set)
```
Comment (MadLittleMods) [inline synapse/types/handlers/sliding_sync.py:503]: Just curious, why prefer this syntax over the `.copy()` method?

File: synapse/types/handlers/sliding_sync.py [hunk 4]
```diff
@@ -159,14 +511,20 @@ def deep_copy(self) -> "RoomSyncConfig":
...
         superset union of the two.
         """
+        timeline_limit = self.timeline_limit
+        required_state_map = {
+            event_type: set(state_keys)
+            for event_type, state_keys in self.required_state_map.items()
+        }
+
         # Take the highest timeline limit
         if self.timeline_limit < other_room_sync_config.timeline_limit:
```
Comment (MadLittleMods) [inline synapse/types/handlers/sliding_sync.py:526]: To get rid of the `self` usage in the body after we assign everything at the top:

```suggestion
        if timeline_limit < other_room_sync_config.timeline_limit:
```

File: synapse/types/handlers/sliding_sync.py [hunk 5]
```diff
@@ -499,7 +500,7 @@ def from_room_config(
 
     def deep_copy(self) -> "RoomSyncConfig":
```
Comment (MadLittleMods) [inline synapse/types/handlers/sliding_sync.py:501]: Since this is immutable now, I don't think we need this function anymore

**Review by MadLittleMods — APPROVED** [review]

File: synapse/types/handlers/sliding_sync.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/types/handlers/sliding_sync.py:503]: `required_state_map`  is an `AbstractSet`, and `copy` doesn't exist.

File: scripts-dev/mypy_synapse_plugin.py (hunk 1 above)
Comment (erikjohnston) [reply scripts-dev/mypy_synapse_plugin.py:323]: Alas that doesn't work, as we're an enum value (which isn't an instance of a subclass of enum). Honestly my brain was melting trying to get all this to work before I found the `is_enum` flag

File: scripts-dev/mypy_synapse_plugin.py (hunk 2 above)
Comment (MadLittleMods) [reply scripts-dev/mypy_synapse_plugin.py:324]: Does this work?

```py
        elif rt.type.is_enum:
            for enum_member_name in rt.get_enum_values():
                enum_member_type = rt.type.names[enum_member_name].type
                # Enum members should have types
                assert enum_member_type is not None
                ok, note = is_cacheable(enum_member_type, signature, verbose)
                if not ok:
                    return (
                        False,
                        f"enum member {enum_member_name} value of type {enum_member_type} not cacheable: {note}",
                    )

            # All Enum member values are immutable
            return True, None
```

It hasn't been tested as this code doesn't seem to matter anymore in this PR? I commented this whole enum logic branch out and `mypy` still passes.

Comment (erikjohnston) [reply scripts-dev/mypy_synapse_plugin.py:324]: I had no idea that get enum values was a thing. I'll have a try and see if it works! Very good catch either way 

Yeah, turns out this matters for the next PR where we add a cache on this value. Not sure why I added this to this pr.

**Pull request merged**